packages/fileeditor/src/toc/markdown.ts

```diff
@@ -34,30 +46,40 @@ export class MarkdownTableOfContentsModel extends TableOfContentsModel<
    *
    * @returns The list of new headings or `null` if nothing needs to be updated.
    */
-  protected getHeadings(): Promise<IEditorHeading[] | null> {
+  protected async getHeadings(): Promise<IEditorHeading[] | null> {
     if (!this.isActive) {
       return Promise.resolve(null);
     }
 
     const content = this.widget.content.model.sharedModel.getSource();
 
     const headings = TableOfContentsUtils.filterHeadings(
-      TableOfContentsUtils.Markdown.getHeadings(content),
+      await TableOfContentsUtils.Markdown.getHeadings(content, this.parser),
       {
         ...this.configuration,
         // Force removing numbering as they cannot be displayed
         // in the document
         numberHeaders: false
       }
     );
-    return Promise.resolve(headings);
+    return headings;
   }
 }
 
 /**
  * Table of content model factory for Markdown files.
  */
 export class MarkdownTableOfContentsFactory extends EditorTableOfContentsFactory {
+  parser: IMarkdownParser | null;
+
+  constructor(
+    tracker: WidgetTracker<IDocumentWidget<FileEditor>>,
+    parser: IMarkdownParser | null
+  ) {
+    super(tracker);
+    this.parser = parser;
+  }
```

Comment:
Can we make it protected?

```suggestion
  constructor(
    tracker: WidgetTracker<IDocumentWidget<FileEditor>>,
    protected parser: IMarkdownParser | null
  ) {
    super(tracker);
  }
```